pyqir-generator/src/python.rs

```diff
@@ -1,174 +1,264 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
-use crate::emit::{get_bitcode_base64_string, get_ir_string, write_model_to_file};
-use crate::interop::{
-    ClassicalRegister, Controlled, Instruction, Measured, QuantumRegister, Rotated, SemanticModel,
-    Single,
+
+use crate::{
+    emit,
+    interop::{
+        ClassicalRegister, Controlled, If, Instruction, Measured, QuantumRegister, Rotated,
+        SemanticModel, Single,
+    },
 };
-use log;
-use pyo3::exceptions::PyOSError;
-use pyo3::prelude::*;
+use pyo3::{exceptions::PyOSError, prelude::*};
 
 #[pymodule]
-fn pyqir_generator(_py: Python<'_>, m: &PyModule) -> PyResult<()> {
-    m.add_class::<PyQIR>()?;
-    Ok(())
+fn _native(_py: Python, m: &PyModule) -> PyResult<()> {
+    m.add_class::<SimpleModule>()?;
+    m.add_class::<Qubit>()?;
+    m.add_class::<Ref>()?;
+    m.add_class::<Builder>()?;
+    m.add_class::<BasicQisBuilder>()
 }
 
+const RESULT_NAME: &str = "result";
+const QUBIT_NAME: &str = "qubit";
+
 #[pyclass]
-pub struct PyQIR {
-    pub(super) model: SemanticModel,
+struct SimpleModule {
+    model: SemanticModel,
+    builder: Py<Builder>,
 }
 
 #[pymethods]
-impl PyQIR {
+impl SimpleModule {
     #[new]
-    fn new(name: String) -> Self {
-        PyQIR {
-            model: SemanticModel::new(name),
+    fn new(py: Python, name: String, num_qubits: u64, num_results: u64) -> PyResult<SimpleModule> {
+        let registers = vec![ClassicalRegister::new(RESULT_NAME.to_string(), num_results)];
+
+        let qubits = (0..num_qubits)
+            .map(|i| QuantumRegister::new(QUBIT_NAME.to_string(), i))
+            .collect();
+
+        let model = SemanticModel {
+            name,
+            registers,
+            qubits,
+            instructions: Vec::new(),
+            static_alloc: true,
+        };
+
+        let builder = Py::new(py, Builder::new())?;
+        Ok(SimpleModule { model, builder })
+    }
+
+    #[getter]
+    fn qubits(&self) -> Vec<Qubit> {
+        self.model
+            .qubits
+            .iter()
+            .map(|q| Qubit { index: q.index })
+            .collect()
+    }
+
+    #[getter]
+    fn results(&self) -> Vec<Ref> {
+        let size = self.model.registers.first().unwrap().size;
+        (0..size)
+            .map(|index| Ref(RefKind::Result { index }))
+            .collect()
+    }
+
+    #[getter]
+    fn builder(&self) -> Py<Builder> {
+        self.builder.clone()
+    }
+
+    fn ir(&mut self, py: Python) -> PyResult<String> {
```

Comment:
Why does this need `&mut self`?

Reply:
Makes sense! As a quick thought should that optimization become necessary again in the future, if the mutation is not externally visible, that may be an appropriate place to use an `Rc<RefCell<_>>` to allow interior mutability without propagating to an observable mutation.